File: typescript/2024/day1.ts
import * as R from 'ramda';

import { parseInt10 } from '../lib/fp';

const content = (await Bun.file('../inputs/2024/Day1/input.txt').text()).trim();

const rLineRegex = /([0-9]+)\s+([0-9]+)/;

const parse = (str: string): [number, number] => {
  const [, a, b] = str.match(rLineRegex)!;
  return [parseInt10(a), parseInt10(b)];
};

const asTwoLists = R.flow(content, [
  c => c.split('\n'),
  c => c.map(parse),
  R.transpose<number>,
  c => c.map(x => x.sort()),
]);

const pairs = R.transpose(asTwoLists);

// console.log(pairs);

const r1 = R.sum(pairs.map(([l, r]) => Math.abs(r - l)));

console.log(r1);

const [ll, rr] = asTwoLists;

const sims = ll.map(v => {
  const otherLength = rr.filter(r => r === v).length;
  return v * otherLength;
});

const r2 = R.sum(sims);
console.log(r2);

File: haskell/2024/Day1.hs
module Day1 where

import Data.List
import Data.Map (Map)
import Data.Map qualified as M
import Data.Maybe

parse :: String -> (Int, Int)
parse s =
  let [(a, s2)] = (reads :: ReadS Int) s
      [(b, _)] = (reads :: ReadS Int) s2
   in (a, b)

main' :: IO ()
main' = do
  contents <- lines <$> readFile "../inputs/2024/Day1/input.txt"
  let pairs = map parse contents

  let (ll, rr) = unzip pairs
  let sorted = zip (sort ll) (sort rr)
  let r1 = sum $ map (abs . uncurry (-)) sorted
  print r1

  let rCounts = foldr (\x -> M.insertWith (+) x 1) M.empty rr
  let r2 = sum $ mapMaybe (\x -> (* x) <$> M.lookup x rCounts) ll
  print r2
